zipline/pipeline/loaders/quarter_estimates.py

```diff
@@ -0,0 +1,391 @@
+                    ][EVENT_DATE_FIELD_NAME],
+                    side=qtr_crossover_point
+                )
+
+                # Only add adjustments if the next quarter starts somewhere in
+                # our date index for this sid. Our 'next' quarter can never
+                # start at index 0; a starting index of 0 means that the next
+                # quarter's event date was NaT.
+                if 0 < qtr_start_idx < len(dates):
+                    # Find the quarter being requested in the quarter we're
+                    # crossing into.
+                    requested_quarter = requested_qtr_data[
+                        SHIFTED_NORMALIZED_QTRS
+                    ][sid].iloc[qtr_start_idx]
+
+                    # If there are estimates for the requested quarter,
+                    # overwrite all values going up to the starting index of
+                    # that quarter with estimates for that quarter.
+                    if requested_quarter in quarters_with_estimates_for_sid:
+                        adjustments[qtr_start_idx] = \
+                            [overwrite(
+                                0,
+                                qtr_start_idx - 1,  # overwrite thru last qtr
+                                sid_idx,
+                                sid_idx,
+                                last_per_qtr[column_name,
+                                             requested_quarter,
+                                             sid][:qtr_start_idx].values)]
+                    # There are no estimates for the quarter. Overwrite all
+                    # values going up to the starting index of that quarter
+                    # with the missing value for this column.
+                    else:
+                        adjustments[qtr_start_idx] = [
+                            overwrite(
+                                0,
+                                qtr_start_idx - 1,
+                                sid_idx,
+                                sid_idx,
+                                np.array(
+                                    [missing_value] *
+                                    len(last_per_qtr.index[:qtr_start_idx]))
+                            )
+                        ]
+
+        return AdjustedArray(
+            requested_qtr_data[column_name].values.astype(column.dtype),
+            mask,
+            dict(adjustments),
+            column.missing_value,
+        )
+
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        # TODO: how can we enforce that datasets have the num_quarters
+        # attribute, given that they're created dynamically?
+        groups = groupby(lambda x: x.dataset.num_quarters, columns)
+        groups_columns = dict(groups)
+        if (np.array(list(groups_columns.keys())) < 0).any():
+            raise ValueError("Must pass a number of quarters >= 0")
+        out = {}
+        self.estimates[NORMALIZED_QUARTERS] = normalize_quarters(
```

Comment:
we don't need to do this each time we call `load_adjusted_array`, could we move this to the constructor?